versal2.mk

```diff
@@ -0,0 +1,173 @@
+# ARM Trusted Firmware
+################################################################################
+
+TF_A_EXPORTS = CROSS_COMPILE="$(CCACHE)$(AARCH64_CROSS_COMPILE)"
+TF_A_FLAGS = PLAT=versal2 CONSOLE=pl011 RESET_TO_BL31=1 SPD=opteed DEBUG=0 \
+	     MEM_BASE=0x1600000 MEM_SIZE=0x200000 \
+	     XILINX_OF_BOARD_DTB_ADDR=0x1000 \
+	     BL32_MEM_BASE=0x1800000 BL32_MEM_SIZE=0x8000000
+
+tfa:
+	$(TF_A_EXPORTS) $(MAKE) -C $(TF_A_PATH) $(TF_A_FLAGS) bl31
+	mkdir -p $(BINARIES_PATH)
+	cp $(TF_A_ELF) $(BINARIES_PATH)
+
+tfa-clean:
+	$(TF_A_EXPORTS) $(MAKE) -C $(TF_A_PATH) $(TF_A_FLAGS) clean
+
+################################################################################
+# OP-TEE
+################################################################################
+
+OPTEE_OS_COMMON_FLAGS += CFG_TEE_CORE_LOG_LEVEL=2 CFG_TEE_TA_LOG_LEVEL=2
+
+optee-os: optee-os-common
+	mkdir -p $(BINARIES_PATH)
+	cp $(OPTEE_OS_ELF) $(BINARIES_PATH)
+
+optee-os-clean: optee-os-clean-common
+	rm -rf ${OPTEE_OS_PATH}/out/
+
+################################################################################
+# U-Boot
+################################################################################
+
+U-BOOT_EXPORTS = CROSS_COMPILE="$(CCACHE)$(AARCH64_CROSS_COMPILE)"
+U-BOOT_DEFCONFIG_COMMON_FILES := $(U-BOOT_PATH)/configs/amd_versal2_virt_defconfig \
+			$(CURDIR)/kconfigs/u-boot_versal2.conf
+
+
+u-boot-defconfig: $(U-BOOT_DEFCONFIG_COMMON_FILES)
+	cd $(U-BOOT_PATH) && \
+                ARCH=arm64 \
+                scripts/kconfig/merge_config.sh $(U-BOOT_DEFCONFIG_COMMON_FILES)
+
+u-boot: u-boot-defconfig
+	$(U-BOOT_EXPORTS) $(MAKE) -C $(U-BOOT_PATH)
+	mkdir -p $(BINARIES_PATH)
+	cp $(U-BOOT_ELF) $(BINARIES_PATH)
+
+u-boot-defconfig-clean:
+	rm -f $(U-BOOT_PATH)/.config
+
+u-boot-clean: u-boot-defconfig-clean
+	$(U-BOOT_EXPORTS) $(MAKE) -C $(U-BOOT_PATH) clean
+
+###############################################################################
+# Device-Tree
+###############################################################################
+dtbo: u-boot
+	mkdir -p $(BINARIES_PATH)
+	${LINUX_PATH}/scripts/dtc/dtc -@ -I dts \
+		-O dtb -o $(BINARIES_PATH)/versal2-memory-reservation.dtbo \
+		./versal2/versal2-memory-reservation.dtso
+	@$(foreach dtb,$(wildcard $(U-BOOT_DTB)), \
+		${LINUX_PATH}/scripts/dtc/fdtoverlay -i $(dtb) \
+		-o $(dtb) $(BINARIES_PATH)/versal2-memory-reservation.dtbo ; \
+		echo "Applied overlay to $(dtb)";)
+	cp $(U-BOOT_DTB) $(BINARIES_PATH)
+
+dtbo-clean:
+	rm -f $(BINARIES_PATH)/versal2-memory-reservation.dtbo
+
+################################################################################
+# Linux kernel
+################################################################################
+
+LINUX_DEFCONFIG_COMMON_ARCH := arm64
+LINUX_DEFCONFIG_COMMON_FILES := \
+		$(LINUX_PATH)/arch/arm64/configs/xilinx_defconfig \
+		$(CURDIR)/kconfigs/versal2.conf
+
+linux-defconfig: $(LINUX_PATH)/.config
+
+LINUX_COMMON_FLAGS += ARCH=arm64 -j8
```

Comment:
`-j$(nproc)` would be better. Or maybe let caller decide how many CPU shall participate to the build.